src/http/request.cpp

```diff
@@ -0,0 +1,117 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "src/http/request.hpp"
+
+namespace mk {
+namespace http {
+
+using namespace mk::net;
+
+void request_connect(Settings settings, RequestConnectCb cb,
+        Poller *poller, Logger *logger) {
+    if (settings.find("url") == settings.end()) {
+        cb(GenericError(), nullptr);
+        return;
+    }
+    ErrorOr<Url> url = parse_url_noexcept(settings.at("url"));
+    if (!url) {
+        cb(url.as_error(), nullptr);
+        return;
+    }
+    if (url->schema == "httpo") {
+        // tor_socks_port takes precedence because it's more specific
+        if (settings.find("tor_socks_port") != settings.end()) {
+            std::string proxy = "127.0.0.1:";
+            proxy += settings["tor_socks_port"];
+            settings["socks5_proxy"] = proxy;
+        } else if (settings.find("socks5_proxy") == settings.end()) {
+            settings["socks5_proxy"] = "127.0.0.1:9050";
+        }
+    }
+    connect(url->address, url->port, cb, settings, logger, poller);
+}
+
+void request_send(Var<Transport> transport, Settings settings, Headers headers,
+        std::string body, RequestSendCb callback) {
+    RequestSerializer serializer;
+    try {
+        serializer = RequestSerializer(settings, headers, body);
+    } catch (std::exception &) {
+        callback(GenericError());
+        return;
+    }
+    transport->on_error([transport, callback](Error error) {
+        transport->on_error(nullptr);
+        transport->on_flush(nullptr);
+        callback(error);
+    });
+    transport->on_flush([transport, callback]() {
+        transport->on_error(nullptr);
+        transport->on_flush(nullptr);
+        callback(NoError());
+    });
+    Buffer buff;
+    serializer.serialize(buff);
+    transport->write(buff);
+}
+
+void request_recv_response(Var<Transport> transport, RequestRecvResponseCb cb,
+        Poller *poller, Logger *logger) {
+    Var<ResponseParserNg> parser(new ResponseParserNg);
+    Var<Response> response(new Response);
+    Var<bool> prevent_emit(new bool(false));
+
+    transport->on_data([=](Buffer data) { parser->feed(data); });
```

Comment:
> I think here you can give to the lambda just the variables it needs!

I prefer not doing that. In my opinion, having verified that it's safe to share all the variables through `[=]`, the safest thing is to always use `[=]` in the closure, so that everything is always shared.

This style, in my opinion, makes it simpler to reason about the code than when you selectively indicate what yo share, because you have a closure of the relevant variables.

I also understand that creating closures like this causes copies that may have impact on performance, but honestly I'd rather measure that we have a problem before starting to optimize.

As for the possibility of sharing pointers that one should not share using `[=]`, this is true but in our case it does not apply. We could use `Var<Poller>` and `Var<Logger> in the future to reduce the theoretical risk of this issue and be more clean. But, again, before going down that road I'd rather do some more experience points trying to use closures.

> should you set the callback to null also here?

The transport's callbacks are cleared as part of the handling of the error event. The error event should be emitted on network error, or when the parser reaches end (with a boolean value protecting double invocation of the error event). We should reason a bit more on whether this would be enough to break possible reference cycles (which is something to be very careful about until we properly hold a reference to transport instances in use in transport code). I will add a comment to remind us to check whether we need to do more.